fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java

```diff
@@ -800,33 +802,76 @@ public boolean addManually(String data) {
    * @return true if an attendee was added false otherwise
    */
   private boolean handleAttendeeAddition(String data) {
+    Log.d(TAG, "data scanned " + data);
+    if (scanningAction == ScanningAction.ADD_ROLL_CALL_ATTENDEE) {
+      return handleRollCallAddition(data);
+    } else if (scanningAction == ScanningAction.ADD_WITNESS) {
+      return handleWitnessAddition(data);
+    } else {
+      throw new IllegalStateException(
+          "The scanning action should either be to add witnesses or rc attendees");
+    }
+  }
+
+  public boolean handleRollCallAddition(String data) {
+    PopTokenData tokenData;
+    try {
+      tokenData = PopTokenData.extractFrom(gson, data);
+    } catch (Exception e) {
+      ErrorUtils.logAndShow(
+          getApplication().getApplicationContext(), TAG, R.string.qr_code_not_pop_token);
+      return false;
+    }
     PublicKey publicKey;
     try {
-      publicKey = new PublicKey(data);
+      publicKey = new PublicKey(tokenData.getPopToken());
     } catch (IllegalArgumentException e) {
       mScanWarningEvent.postValue(new SingleEvent<>("Invalid key format code. Please try again."));
       return false;
     }
-    if (attendees.contains(publicKey) || witnesses.contains(publicKey)) {
+    if (attendees.contains(publicKey)) {
       mScanWarningEvent.postValue(
           new SingleEvent<>("This attendee key has already been scanned. Please try again."));
       return false;
     }
-    if (scanningAction == ScanningAction.ADD_ROLL_CALL_ATTENDEE) {
-      attendees.add(publicKey);
-      mAttendeeScanConfirmEvent.postValue(new SingleEvent<>("Attendee has been added."));
-      mNbAttendees.postValue(attendees.size());
-    } else if (scanningAction == ScanningAction.ADD_WITNESS) {
-      witnesses.add(publicKey);
-      mWitnessScanConfirmEvent.postValue(new SingleEvent<>(true));
-      disposables.add(
-          updateLaoWitnesses()
-              .subscribe(
-                  () -> Log.d(TAG, "Witness " + publicKey + " added"),
-                  error ->
-                      ErrorUtils.logAndShow(
-                          getApplication(), TAG, error, R.string.error_update_lao)));
+    attendees.add(publicKey);
+    mAttendeeScanConfirmEvent.postValue(new SingleEvent<>("Attendee has been added."));
+    mNbAttendees.postValue(attendees.size());
+    return true;
+  }
+
+  public boolean handleWitnessAddition(String data) {
```

Comment:
fe-1 does not have an analogous method 🙃 Is this usable or just here to support future extensions?

Reply:
Yeah we can add witnesses but I don't if it's then possible to have them something useful

Reply:
They should be able to sign messages in the consensus